lib/validator.js

```diff
@@ -34,16 +35,29 @@ async function validateExpectations() {
 		throw new Error("requiredExpectaions should be of type Set");
 	}
 
+	if (Array.isArray(opt)) {
+		opt = new Set([opt]);
+	}
+
+	if (!(opt instanceof Set)) {
+		throw new Error("optionalExpectations should be of type Set");
+	}
+
 	for (let field of req) {
 		if (!exp.has(field)) {
 			throw new Error(`expectation did not contain value for '${field}'`);
 		}
 	}
 
-	let expSize = exp.size;
-	if (exp.has("rpId")) expSize--; // don't include optional fields
-	if (req.size !== expSize) {
-		throw new Error(`wrong number of expectations: should have ${req.size} but got ${exp.size}`);
+	let optCount = 0;
+	for (const [field] of exp) {
+		if (opt.has(field)) {
+			optCount++;
+		}
+	}
+
+	if (req.size !== exp.size - optCount) {
```

Comment:
Wouldnt it make more sense to count the mandatory fields and compare it without the math?

Reply:
Sure, what I meant was that I think it would make more sense to count the found mandatory fields in the loop above and then check if this number matches - this is easier to understand and has less potential side effects

Reply:
yep but in that case we wouldn't need to compare counts since it wouldn't reach here at all if at least one required field isn't found due to thrown error (if (!exp.has(field)) throw new Error(`expectation did not contain value for '${field}'`)).

also counting required fields wouldn't check for other fields that were provided but are unknown (neither optional or required)

Reply:
Hi @JamesCullum, could you please check the comment above?